File: tools/xread/tests/fixtures/sample.ts
// Sample TypeScript module for xread tests.

import { EventEmitter } from "events";

export interface Greetable {
  name: string;
  greet(): string;
}

export type Language = "en" | "es";

// Format a name for display.
export function formatName(first: string, last: string): string {
  return `${first} ${last}`;
}

export const shout = (text: string): string => {
  return text.toUpperCase() + "!";
};

export class Greeter extends EventEmitter {
  private language: Language;

  constructor(language: Language = "en") {
    super();
    this.language = language;
  }

  // The main entry point for greetings.
  greet(target: Greetable): string {
    if (this.language === "en") {
      return `hello, ${target.name}`;
    }
    return `hola, ${target.name}`;
  }

  static default(): Greeter {
    return new Greeter();
  }
}

export default function farewell(name: string): string {
  return `bye, ${name}`;
}

File: tools/xread/tests/fixtures/sample.py
"""Sample module for xread tests."""

import math

GREETING = "hello"


# Compute the hypotenuse of a right triangle.
# Used by the query-mode tests.
def hypot(a, b):
    """Return the hypotenuse length."""
    return math.sqrt(a * a + b * b)


def _cache(func):
    """Tiny memoizing decorator."""
    seen = {}

    def wrapper(*args):
        if args not in seen:
            seen[args] = func(*args)
        return seen[args]

    return wrapper


@_cache
def slow_square(n):
    """Square a number, memoized."""
    return n * n


class Greeter:
    """Greets people in a configurable language."""

    def __init__(self, language="en"):
        self.language = language

    # The main entry point for greetings.
    @_cache
    def greet(self, name):
        """Return a greeting for name."""
        if self.language == "en":
            return "%s, %s!" % (GREETING, name)
        return "hola, %s!" % name

    async def greet_async(self, name):
        return self.greet(name)


def farewell(name):
    """Say goodbye."""
    return "bye, %s" % name
